Enable `dim_order` kwargs for direct image conversions

elevans opened the pull request:
This pull request enables `dim_order` kwargs for the direct image converter methods: `to_dataset()`, `to_img()`, and `to_xarray()`. It also adds tests for this new feature in `test_image_conversion.py` and some additional features. There are also a couple of bugs/enhancements I made (_e.g._ we now calculate the slope for linear axis using all points instead of the first two elements of the scale array).

To get a good idea on the `dim_order` kwarg option and how it works with the direct image converters, check out the new section I wrote in the docs (`06-Working-with-Images`, section 6.9). There is a nice table that outlines the image conversion behavior with or without `dim_order` supplied.

## changes

- Enable `dim_order` kwarg for direct image converions (`to_dataset()`, `to_img()`, `to_xarray()`).
- Add tests for `dim_order` kwargs
- Refactor image tests to use pytest's parameterize syntax.
- Add test to check dimension coordinates are preserved between image conversions.
- ~~Replaced `imagej.dims._get_scale()` with `imagej.dims._compute_slope()`.~~ (see latest comment)
- Allow for `dim_order` lengths that are shorter than the length of the array. Unknown dims are assigned `dim_n` like xarray.
- Add 6.9 -- information about using the direct image converters and `dim_order`.
- Add singleton scale hack.

gselzer commented on src/imagej/dims.py:
```diff
@@ -299,19 +284,56 @@ def _get_enumerated_axis():
 
+def _compute_slope(axis):
+    """
+    Compute slope (i.e. scale) of an axis with all points.
+
```
I can't say I really understand what this means. What slope is calculated? Can we elaborate a bit in this docstring?

elevans replied:
Yes, I agree this is nebulous. I'll improve this docstring. But to briefly summarize what's going on here:
 
Axes have coordinates along each axis (a 1D array) which can be linear or non-linear. When a user converts an `xarray.DataArray` with coordinates. Newer versions of ImageJ come with `EnumeratedAxis` which makes converting the coordinates really easy and also handles non-linear coordinates. If `EnumeratedAxis` is not available however, we fall back to `DefaultLinearAxis` which requires the slope of the 1D coordinate array. Prior to this PR we calculated the slope by subtracting the first two elements from each other from the coordinate array. I originally had implemented linear regression so we could actually decide if coordinates were linear or not. But that was kind of silly and not really necessary. Part of that however, involved calculating the slope using all points of the array instead of just the first two elements. I thought this was a more robust slope calculation and got values that matched ImageJ's slope values (`float`).

gselzer replied:
Gotcha, thanks for the explanation. Having this explanation in the docstring would be great!

elevans replied:
Improved the docstring: https://github.com/imagej/pyimagej/pull/238/commits/0126d0039cc2a6bc7b5531bce19a2c2e28825790

gselzer reviewed (approved):
I didn't take the time to understand to understand the dimension translations, but this looks like a great step forward @elevans!

Just had a couple minor suggestions, but I don't think anything was major.

elevans commented on src/imagej/dims.py:
```diff
@@ -198,26 +198,26 @@ def _assign_axes(xarr: xr.DataArray):
 
-        if scale is None:
+        if _is_numeric_scale(scale_arr):
+            doub_coords = [Double(np.double(x)) for x in xarr.coords[dim]]
```
I was thinking the same thing here as well. But when I was looking at what classes we have loaded it felt kind of ImageJ specific and adding a Java primitive felt a bit weird. Thoughts? Should we just add all the primitive types there?

gselzer replied:
I use the `JavaClasses` object in napari-imagej to be DRY, and to put all of the `jimport` calls in the same place. It's up to you whether you want to do the same, but I'd advise it if you use `Double` anywhere else.

elevans replied:
I agree! https://github.com/imagej/pyimagej/pull/238/commits/0479a3cf9e822d03227d176be93db8a6b1f8189a

elevans commented on src/imagej/dims.py:
```diff
@@ -198,26 +198,26 @@ def _assign_axes(xarr: xr.DataArray):
         # EnumeratedAxis is a new axis made for xarray, so is only present in
         # ImageJ versions that are released later than March 2020.
         # This actually returns a LinearAxis if using an earlier version.
         if EnumeratedAxis is not None:
```
I agree! https://github.com/imagej/pyimagej/pull/238/commits/41d5851d4be8e0d623336b2ea4bf45d5927b1a75

elevans commented on tests/test_image_conversion.py:
```diff
@@ -6,115 +6,108 @@
+    for i in range(len(dims)):
+        dims[i] = i + 1
+    ns = ij_fixture.op().namespace(CreateNamespace)
+    img = ns.img(dims)
```
Yeah it looks like this works! The reason its currently done with `CreateNamespace`  is because of the `ij.op().create()` ambiguity issue (https://github.com/imagej/pyimagej/issues/106).

I like your suggestion more. It is easier to read/understand. https://github.com/imagej/pyimagej/commit/ea9fc13c4c5b83149ad5715eee366a055120b69c

gselzer replied:
Gotcha, good to know that it's completed and that we can write something simpler!

elevans commented:
## Update:

I'm working on using xarray's global attribute's to store imagej/pyimagej specific metadata like the scale and `CalibratedAxis` type. See the [axis-scale-logic](https://github.com/imagej/pyimagej/tree/axis-scale-logic) branch for more details. The way things are going it seems this will be our preferred way to assign `CalibratedAxis` types. Thus I think we should scrap the scale work I've done here (_i.e._ `_compute_scale()` _etc_...).

elevans commented:
Me again! I have reverted the `dims.py` module to a near original state. The difference from `main` is:
- Added `_validate_dim_order()` function which allows users to pass dimension lists that are smaller than the size of the data. The unknown dimensions are then populated with `dim_n` where `n` is incrementing. This allows us to label unknown dimensions in the same way as xarray.
- Added a hack to deal with converting `numpy.ndarrays` or `xarray.DataArrays` with a singleton dimension. If a singleton dimension is in the data we simply assign scale=1 for the axis. Without this hack you hit an index error when trying to do the scale math. See below for more details.

I have a nice metadata solution and a rework on how we assign axes for xarrays <-> datasets. That work is still in progress (almost done) on this branch/PR: https://github.com/imagej/pyimagej/pull/247

## Singleton dimensions

The reason I originally added the work with axis scales was because I discovered a bug when creating more tests for the `dim_order` kwargs. When converting a `numpy.ndarray` or an `xarray.DataArray` that has a singleton dimension (_i.e._ one or more axes is size 1) we run into an index error when trying to perform a subtraction between the first and second elements of the axis/scale array here: https://github.com/imagej/pyimagej/blob/a5544fa07a1b3652c2c0037afbbd07e0f5d32add/src/imagej/dims.py#L277-L289

From what I gather rather serious bug has been here the entire time. Here is how you can reproduce it on `main`:

```python
import numpy as np
import imagej

ij = imagej.init()
nparr = np.random.rand(1, 2, 3, 4, 5)

# index error
ds = ij.py.to_dataset(nparr)
```

I think no one has ran into this yet (or very few people have) because when you slice an array to a singleton dimension its usually squashed. But anyone who is making an array with a singleton dimension will hit this bug when converting to a `Dataset`.
